cpp/src/io/csv/writer_impl.cu

```diff
@@ -0,0 +1,611 @@
+              std::ostream_iterator<std::string>(ss, delimiter_str.c_str()));
+    ss << metadata->column_names.back();
+
+    out_sink_->host_write(ss.str().data(), ss.str().size());
+  }
+}
+
+void writer::impl::write_chunked(strings_column_view const& strings_column,
+                                 const table_metadata* metadata,
+                                 cudaStream_t stream)
+{
+  // algorithm outline:
+  //
+  //  for_each(strings_column.begin(), strings_column.end(),
+  //           [sink = out_sink_](auto str_row) mutable {
+  //               auto host_buffer = str_row.host_buffer();
+  //               sink->host_write(host_buffer_.data(), host_buffer_.size());
+  //           });//or...sink->device_write(device_buffer,...);
+
+  auto pair_buff_offsets = cudf::strings::create_offsets(strings_column, stream, mr_);
+
+  auto num_rows                  = strings_column.size();
+  decltype(num_rows) num_offsets = pair_buff_offsets.second.size();
+
+  CUDF_EXPECTS(num_rows == num_offsets,
+               "Unexpected discrepancy between number of offsets and number of rows.");
+
+  auto total_num_bytes      = pair_buff_offsets.first.size();
+  char const* ptr_all_bytes = pair_buff_offsets.first.data().get();
+
+  rmm::device_vector<size_type> d_row_sizes(num_rows);
+
+  // extended device lambdas called inside a member function
+  // of a nested classes need:
+  //(1) the nested class to be declared public;
+  //(2) the calling member function to be public;
+  //(otherwise known compiler error)
+  //
+  // extract row sizes in bytes from the offsets:
+  //
+  auto exec = rmm::exec_policy(stream);
+  thrust::transform(exec->on(stream),
+                    thrust::make_counting_iterator<size_type>(0),
+                    thrust::make_counting_iterator<size_type>(num_offsets),
+                    d_row_sizes.begin(),
+                    [ptr_all_bytes, total_num_bytes, num_offsets] __device__(auto row_index) {
+                      return row_index < num_offsets - 1
+                               ? ptr_all_bytes[row_index + 1] - ptr_all_bytes[row_index]
+                               : total_num_bytes - ptr_all_bytes[row_index];
+                    });
+
+  // copy offsets to host:
+  //
+  thrust::host_vector<size_type> h_offsets(num_offsets);
+  CUDA_TRY(cudaMemcpyAsync(h_offsets.data(),
+                           pair_buff_offsets.second.data().get(),
+                           num_offsets * sizeof(size_type),
+                           cudaMemcpyDeviceToHost,
+                           stream));
+
+  // copy sizes to host:
+  //
+  thrust::host_vector<size_type> h_row_sizes(num_rows);
+  CUDA_TRY(cudaMemcpyAsync(h_row_sizes.data(),
+                           d_row_sizes.data().get(),
+                           num_rows * sizeof(size_type),
+                           cudaMemcpyDeviceToHost,
+                           stream));
+
+  CUDA_TRY(cudaStreamSynchronize(stream));
+
```

Comment:
If I understand how this works, it is copying the string view's offsets/sizes to host to then write the output rows one row at a time ?
Wouldn't it be possible to just write out the string column's character data buffer ?

Reply:
Not sure what solution you're suggesting here (perhaps a pseudocode sketch might work), but the API that I have available to write into the `sink` is `sink->device_write(ptr, sz, stream); ` (which is a _host_ function, hence moving some data -not all- on host). 

Psudocode of what is happening in that code:
```
pair(d_chars, d_row_offsets) d_pair = create_offsets(...);
h_row_offsets = copy_to_host(d_row_offsets);
h_sizes = compute_sizes_from_offsets(d_row_offsets);
traverse_rows_one_at_a_time(counter(i), d_chars, h_row_offsets, h_sizes, 
                                               sink->device_write(d_chars + h_row_offsets[i], h_sizes[i], stream));
```
Now, I suspect you meant: can't I just do `sink->device_write(d_chars, total_sz_all_rows, stream)`, right? 

Well:

1. I realize `total_num_bytes` is wrong, it's just the number of offsets not the cumulative size in bytes of all rows (good catch, thanks!);
2. If I can get `total_num_bytes` somehow, then I suspect you can, since `d_chars` is a contiguous buffer of all characters in all rows (one after the other). I just need to get the its size in bytes, perhaps by traversing the `strings_column_view` and accumulating (reduce(+)) all its `string_view::size_bytes()` (unless there's a more efficient way to do it).

Am I understanding your comment correctly?

Reply:
Correct. There should be an easy way to get the size of the character buffer, worst case we coould just use `d_offsets[num_rows] - d_offsets[0]`, since there are `numrows+1` offsets (though I think the offset of the first string must be zero), but looking at the header, it looks like there is no need for any of that and that's exactly what `string_column_view::chars_size()` is intended for

Reply:
Fixed.